java/ql/src/utils/stub-generator/Stubs.qll

```diff
@@ -404,20 +400,45 @@ private string stubMember(Member m) {
 
 private string stubAnnotation(Annotation a) {
   if exists(a.getAValue())
-  then result = a.toString() + "(" + concat(stubAnnotationValue(a.getAValue()), ",") + ")"
-  else result = a.toString()
+  then
+    result =
+      a.getType().getName() + "(" +
+        concat(string name, Expr value |
+          value = a.getValue(name)
+        |
+          name + "=" + stubAnnotationValue(value), ","
+        ) + ")"
+  else result = a.getType().getName()
 }
 
-private string stubAnnotiationSimpleValue(Expr value) {
-  result = value.(FieldAccess).getField().getQualifiedName() or
-  result = value.(Literal).toString()
+private string stubAnnotationSimpleValue(Expr value) {
+  result = value.(FieldAccess).getField().getQualifiedName()
+  or
+  result = value.(Literal).getLiteral()
+  or
+  not value instanceof Literal and
+  (
+    result = value.(CompileTimeConstantExpr).getStringValue() or
+    result = value.(CompileTimeConstantExpr).getBooleanValue().toString() or
+    result = value.(CompileTimeConstantExpr).getIntValue().toString()
+  )
+  or
+  result = value.(Annotation).getType().getName()
+  or
+  result = value.(TypeLiteral).getReferencedType().getName() + ".class"
 }
 
 private string stubAnnotationValue(Expr value) {
-  result = stubAnnotiationSimpleValue(value)
+  result = stubAnnotationSimpleValue(value)
   or
   value instanceof ArrayInit and
-  result = "{" + concat(stubAnnotiationSimpleValue(value.(ArrayInit).getAnInit()), ",") + "}"
+  result =
+    "{" +
+      concat(int i, Expr arrayElement |
+        arrayElement = value.(ArrayInit).getInit(i)
```

Comment:
Might be necessary to verify that this behaves as expected, and does not include duplicate values due to #8645.

Reply:
Makes sense. I added `i >= 0`, which should handle this issue just in case, thanks!